subcommands/keys/rotate_root.go

```diff
@@ -0,0 +1,290 @@
+
+	fmt.Println("= Resigning root.json")
+	signers := []TufSigner{
+		{Id: curid, Key: curPk},
+		{Id: newid, Key: newPk},
+	}
+	subcommands.DieNotNil(signRoot(root, signers...))
+
+	bytes, err := json.Marshal(root)
+	subcommands.DieNotNil(err)
+
+	// Create a backup before we try and commit this:
+	tmpCreds := saveTempCreds(credsFile, newCreds)
+
+	fmt.Println("= Uploading rotated root")
+	body, err := api.TufRootPost(factory, bytes)
+	if err != nil {
+		fmt.Println("\nERROR:", err)
+		fmt.Println(body)
+		fmt.Println("A temporary copy of the new root was saved:", tmpCreds)
+		fmt.Println("Before deleting this please ensure your factory isn't configured with this new key")
+		os.Exit(1)
+	}
+	if err := os.Rename(tmpCreds, credsFile); err != nil {
+		fmt.Println("\nERROR: Unable to update offline creds file.", err)
+		fmt.Println("Temp copy still available at:", tmpCreds)
+		fmt.Println("This temp file contains your new factory root private key. You must copy this file.")
+	}
+}
+
+func removeUnusedKeys(root *client.AtsTufRoot) {
+	var inuse []string
+	for _, role := range root.Signed.Roles {
+		inuse = append(inuse, role.KeyIDs...)
+	}
+	// we also have to be careful to not loose the extra root key when doing
+	// a root key rotation
+	for _, sig := range root.Signatures {
+		inuse = append(inuse, sig.KeyID)
+	}
+
+	for k := range root.Signed.Keys {
+		// is k in inuse?
+		found := false
+		for _, val := range inuse {
+			if k == val {
+				found = true
+				break
+			}
+		}
+		if !found {
+			fmt.Println("= Removing unused key:", k)
+			delete(root.Signed.Keys, k)
+		}
+	}
+}
+
+func signRoot(root *client.AtsTufRoot, signers ...TufSigner) error {
+	removeUnusedKeys(root)
+
+	bytes, err := canonical.MarshalCanonical(root.Signed)
+	if err != nil {
+		return err
+	}
+
+	opts := rsa.PSSOptions{SaltLength: 32, Hash: crypto.SHA256}
+	hashed := sha256.Sum256(bytes)
+
+	root.Signatures = []tuf.Signature{}
+
+	for _, signer := range signers {
+		bytes, err = signer.Key.Sign(rand.Reader, hashed[:], &opts)
+		if err != nil {
+			return err
+		}
+		sig := tuf.Signature{
+			KeyID:     signer.Id,
+			Method:    "rsassa-pss-sha256",
+			Signature: bytes,
+		}
+		root.Signatures = append(root.Signatures, sig)
+	}
+	return nil
+}
+
+func swapRootKey(root *client.AtsTufRoot, curid string, creds OfflineCreds) (string, *rsa.PrivateKey, OfflineCreds) {
+	pk, err := rsa.GenerateKey(rand.Reader, 2048)
+	subcommands.DieNotNil(err)
+
+	var privBytes []byte = x509.MarshalPKCS1PrivateKey(pk)
+	block := &pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: privBytes,
+	}
+	privBytes, err = json.Marshal(client.AtsKey{
+		KeyType:  "RSA",
+		KeyValue: client.AtsKeyVal{Private: string(pem.EncodeToMemory(block))},
+	})
+	subcommands.DieNotNil(err)
+
+	pubBytes, err := x509.MarshalPKIXPublicKey(&pk.PublicKey)
+	subcommands.DieNotNil(err)
+
+	block = &pem.Block{
+		Type:  "PUBLIC KEY",
+		Bytes: pubBytes,
+	}
+	id := fmt.Sprintf("%x", sha256.Sum256(privBytes))
```

Comment:
Can you put a comment why we generate an ID from privBytes instead of a canonical form of a AtsKey (as prescribed by a TUF spec) ?
Also, did you verify that this is really what aktualizr requires ?

Reply:
An again, generating from privBytes is a security hole, it should be pubBytes in any case.

Reply:
it was just missed when reshuffling. https://github.com/foundriesio/fioctl/pull/83/commits/c8e1ffb3e85fcee03282bad2f65058c5d76b1075

Reply:
ok, but why do we create an ID off a key, and not a key JSON (as TUF spec advices)?

What I mean is this:
```
	pubKey = client.AtsKey{
		KeyType:  "RSA",
		KeyValue: client.AtsKeyVal{Public: string(pem.EncodeToMemory(block))},
	}
        // In your code below two lines are swapped.
        pubBytes, err = json.Marshal(pubKey)
	id := fmt.Sprintf("%x", sha256.Sum256(pubBytes))
        root.Signed.Keys[id] = pubKey
```

Also, I guess we need to use canonical, not a simple json here.

Reply:
i've just done it because aklite and ota-reposerver don't. I've pushed https://github.com/foundriesio/fioctl/pull/83/commits/9bfd1e0b65610bea959256436a9c89eeed1f134e .